buffer/ring_growing.go

```diff
@@ -72,11 +86,44 @@ func (r *RingGrowing) WriteOne(data interface{}) {
 }
 
 // Len returns the number of items in the buffer.
-func (r *RingGrowing) Len() int {
+func (r *TypedRingGrowing[T]) Len() int {
 	return r.readable
 }
 
 // Cap returns the capacity of the buffer.
-func (r *RingGrowing) Cap() int {
+func (r *TypedRingGrowing[T]) Cap() int {
 	return r.n
 }
+
+// Ring is a dynamically-sized ring buffer which can grow and shrink as-needed.
+// Not thread safe.
+type Ring[T any] struct {
+	TypedRingGrowing[T]
+
+	normalSize int // Limits the size of the buffer that is kept for reuse. Read-only.
```

Comment:
`queue.FIFO` is set up to not require any constructors like this, so users can initialize one directly as the zero value and get a valid, empty queue. One detail I couldn't quite iron out to enable that here is how to make the [`normalSize`](https://github.com/kubernetes/kubernetes/blob/ff25881bc1fe9da1dda4ec82d294b92078ae2763/pkg/scheduler/util/queue/fifo.go#L20-L22) configurable. There it's hardcoded to 4 which seems reasonable for where that's used, but I see informers' notification buffers using `RingGrowing` are initialized to 1024, so shrinking all the way down to 4 could result in significantly more allocations. So if there isn't a single golden value for every use case, I think that value probably has to be fed in through a constructor (or possibly a method).

Reply:
I agree that a constructor parameter would be good. Two int parameters with no clear obvious ordering will be hard to read at call sites, so I prefer the config struct pattern.

We could use optional `With*` methods, but that raises the question whether they return a deep copy of the instance or completely new ones - let's not go there.

Reply:
Regarding the zero value `Ring[T]`: let's document it as `NewRing()` with zero initial size and zero normal size. Perhaps not very useful, but at least it's then documented.

Reply:
Added options structs and mentioned zero values in doc strings.